crates/oxc_linter/src/rules/react/jsx_key.rs

```diff
@@ -0,0 +1,395 @@
+use oxc_ast::{
+    ast::{
+        ArrayExpressionElement, Expression, JSXAttributeItem, JSXAttributeName, JSXElement,
+        JSXFragment,
+    },
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+enum JsxKeyDiagnostic {
+    #[error("eslint-plugin-react/jsx-key: Missing \"key\" prop for element in array.")]
+    #[diagnostic(severity(warning), help("Missing \"key\" prop for element in array."))]
+    MissingKeyPropForElementInArray(#[label] Span),
+    #[error("eslint-plugin-react/jsx-key: Missing \"key\" prop for element in iterator.")]
+    #[diagnostic(severity(warning), help("Missing \"key\" prop for element in iterator."))]
+    MissingKeyPropForElementInIterator(#[label] Span),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct JsxKey;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Enforce `key` prop for elements in array
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// [1, 2, 3].map(x => <App />);
+    /// [1, 2, 3]?.map(x => <BabelEslintApp />)
+    ///
+    /// // Good
+    /// [1, 2, 3].map(x => <App key={x} />);
+    /// [1, 2, 3]?.map(x => <BabelEslintApp key={x} />)
+    /// ```
+    JsxKey,
+    correctness
+);
+
+impl Rule for JsxKey {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::ArrayExpressionElement(ArrayExpressionElement::Expression(expr)) => match expr
+            {
+                Expression::JSXElement(jsx_elem) => {
+                    check_jsx_element(node, jsx_elem, ctx);
+                }
+                Expression::JSXFragment(jsx_frag) => {
+                    check_jsx_fragment(node, jsx_frag, ctx);
+                }
+                _ => {}
+            },
```

Comment:
Is this redundant since you are already checking for arrays in parent in `is_in_array_or_iter`?

Reply:
ooo thanks for making me look again - guess i was too tired when i was doing this the first time (there were a bunch of duplicate errors reported.

I updated the `visit` trait to enter JSX fragmetns and that simplifed it significantly (before we just visited the children of the fragment